// @sudoku/stores/candidates.js
import { writable } from 'svelte/store';

/**
 * 把界面坐标转换成 store 中使用的 key。
 *
 * @param {{x: number, y: number}} pos 单元格坐标
 * @returns {string} 例如 "2,0"
 */
function getKey(pos) {
	return pos.x + ',' + pos.y;
}

/**
 * 管理候选数标记的 Svelte store。
 *
 * 旧项目已经有“笔记模式”，HW2 的候选数提示也可以复用同一套显示组件。
 * 因此这里新增 set / clearAll，让领域层算出的候选数可以直接显示到棋盘上。
 */
function createCandidates() {
	const candidates = writable({});

	return {
		subscribe: candidates.subscribe,

		/**
		 * 切换某个候选数。
		 * 如果不存在就添加；如果已经存在就移除。
		 *
		 * @param {{x: number, y: number}} pos 单元格坐标
		 * @param {number} candidate 候选数字
		 */
		add(pos, candidate) {
			candidates.update($candidates => {
				const key = getKey(pos);
				const current = $candidates[key] ? [...$candidates[key]] : [];
				const index = current.indexOf(candidate);

				if (index >= 0) {
					current.splice(index, 1);
				} else {
					current.push(candidate);
					current.sort((a, b) => a - b);
				}

				if (current.length === 0) {
					delete $candidates[key];
				} else {
					$candidates[key] = current;
				}

				return { ...$candidates };
			});
		},

		/**
		 * 直接设置某格的候选数列表。
		 *
		 * @param {{x: number, y: number}} pos 单元格坐标
		 * @param {number[]} candidateList 候选数列表
		 */
		set(pos, candidateList) {
			candidates.update($candidates => {
				const key = getKey(pos);
				const list = [...new Set(candidateList)].sort((a, b) => a - b);

				if (list.length === 0) {
					delete $candidates[key];
				} else {
					$candidates[key] = list;
				}

				return { ...$candidates };
			});
		},

		/**
		 * 清除某格的候选数。
		 *
		 * @param {{x: number, y: number}} pos 单元格坐标
		 */
		clear(pos) {
			candidates.update($candidates => {
				delete $candidates[getKey(pos)];
				return { ...$candidates };
			});
		},

		/** 清除全盘候选数。 */
		clearAll() {
			candidates.set({});
		}
	};
}

export const candidates = createCandidates();
